internal/provider/resource_user.go

```diff
@@ -5,143 +5,242 @@ import (
-
-	var diags diag.Diagnostics
+// Read is called when the provider must read resource values in order
+// to update state. Planned state values should be read from the
+// ReadRequest and new state values set on the ReadResponse.
+// Take the juju api input from the ID, it may not exist in the plan.
+// Only set optional values if they exist.
+func (r *userResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	var data userResourceModel
+
+	// Read Terraform prior state data into the model
+	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
 
-	id := strings.Split(d.Id(), ":")
-	name := id[1]
-	response, err := client.Users.ReadUser(name)
+	// If applicable, this is a great opportunity to initialize any necessary
+	// provider client data and make a call using it.
+	response, err := r.client.Users.ReadUser(userNameFromID(data.ID.ValueString()))
 	if err != nil {
-		return diag.FromErr(err)
+		// TODO (hmlanigan) 2023-06-14
+		// Add a user NotFound error type to the client.
+		// On read, if NotFound, remove the resource:
+		// resp.State.RemoveResource()
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to read user, got error: %s", err))
+		return
 	}
+	tflog.Trace(ctx, fmt.Sprintf("read user resource %q", data.Name.ValueString()))
 
-	if err := d.Set("name", response.UserInfo.Username); err != nil {
-		return diag.FromErr(err)
+	// Save updated data into Terraform state
+	plan := userResourceModel{
+		Name:     types.StringValue(response.UserInfo.Username),
+		Password: data.Password,
+		ID:       types.StringValue(fmt.Sprintf("user:%s", response.UserInfo.Username)),
 	}
-	if err := d.Set("display_name", response.UserInfo.DisplayName); err != nil {
-		return diag.FromErr(err)
+	// Display name is optional, therefore if it doesn't exist in the plan
+	// if not set, do not add an empty string as they are not the same thing.
+	// Conversely, if the returned user info does not contain an empty string,
+	// make it of type ValueStateNull to indicate not set.
+	if !data.DisplayName.IsNull() && !data.DisplayName.IsUnknown() && response.UserInfo.DisplayName != "" {
+		plan.DisplayName = types.StringValue(response.UserInfo.DisplayName)
+	} else if response.UserInfo.DisplayName == "" {
+		plan.DisplayName = types.StringNull()
 	}
-
-	return diags
+	resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
 }
 
-func resourceUserUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	client := meta.(*juju.Client)
-
-	var diags diag.Diagnostics
-	anyChange := false
+// Update is called to update the state of the resource. Config, planned
+// state, and prior state values should be read from the
+// UpdateRequest and new state values set on the UpdateResponse.
+func (r *userResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	var data userResourceModel
```

Comment:
I think I called this `state` to indicate the "current state", whereas I understand the `plan` indicates the desired state. `data` makes sense too, but I'd like to discuss and settle on one term to be consistent across all the resources in the provider.

Reply:
This is what is written in terraform itself.